src/algebra/lie_algebra.lean

```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+⟨lie_subalgebra.to_submodule⟩
+
+instance lie_subalgebra_lie_algebra [L' : lie_subalgebra R L] : lie_algebra R L' := {
+  bracket  := λ x y, ⟨⁅x.val, y.val⁆, L'.bracket x.property y.property⟩,
+  lie_add  := by { intros, apply set_coe.ext, apply lie_add, },
+  add_lie  := by { intros, apply set_coe.ext, apply add_lie, },
+  lie_self := by { intros, apply set_coe.ext, apply lie_self, },
+  jacobi   := by { intros, apply set_coe.ext, apply lie_ring.jacobi, },
+  lie_smul := by { intros, apply set_coe.ext, apply lie_smul, } }
+
+end lie_subalgebra
+
+section lie_module
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+variables (M  : Type v) [add_comm_group M] [module R M]
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A Lie module is a module over a commutative ring, together with a linear action of a Lie algebra
+on this module, such that the Lie bracket acts as the commutator of endomorphisms.
+-/
+class lie_module := (action : L →ₗ⁅R⁆ module.End R M)
+end prio
+
+/--
+Every Lie algebra is a module over itself.
+-/
+instance lie_algebra_self_module : lie_module R L L := ⟨lie_algebra.bil_lie_morphism⟩
+
+/--
+A Lie submodule of a Lie module is a submodule that is closed under the Lie bracket.
+This is a sufficient condition for the subset itself to form a Lie module.
+-/
+structure lie_submodule [lie_module R L M] extends submodule R M :=
+(bracket : ∀ {x m}, m ∈ carrier → lie_module.action R L M x m ∈ carrier)
+
+instance lie_submodule_coe_submodule [lie_module R L M]
+  : has_coe (lie_submodule R L M) (submodule R M) := ⟨lie_submodule.to_submodule⟩
+
+instance lie_submodule_has_mem [lie_module R L M]
+  : has_mem M (lie_submodule R L M) := ⟨λ x M', x ∈ (M' : set M)⟩
+
+instance lie_submodule_lie_module [α : lie_module R L M] [M' : lie_submodule R L M]
+  : lie_module R L M' := ⟨{
+  add     := by { intros, ext, apply set_coe.ext,
+                  rw [linear_map.coe_mk, subtype.coe_mk, linear_map.add], refl, },
+  smul    := by { intros, ext, apply set_coe.ext,
+                  rw [linear_map.coe_mk, subtype.coe_mk, linear_map.smul], refl, },
+  bracket := by { intros, ext, apply set_coe.ext,
+                  rw [linear_map.coe_mk, subtype.coe_mk, lie_algebra.morphism.bracket], refl, },
+  to_fun  := λ x, {
+    to_fun := λ m, ⟨α.action.to_fun x m, M'.bracket m.property⟩,
+    add    := by { intros, apply set_coe.ext, simp, },
+    smul   := by { intros, apply set_coe.ext, simp, }}}⟩
+
+/--
+An ideal of a Lie algebra is a Lie submodule of the Lie algebra as a Lie module over itself.
+-/
+abbreviation lie_ideal := lie_submodule R L L
+
+lemma lie_bracket_mem_right (I : lie_ideal R L) (x y : L) (h : y ∈ I) : ⁅x, y⁆ ∈ I := I.bracket h
+
+lemma lie_bracket_mem_left (I : lie_ideal R L) (x y : L) (h : x ∈ I) : ⁅x, y⁆ ∈ I := by {
+  rw [←lie_skew, ←neg_lie], apply lie_bracket_mem_right, assumption, }
+
```

Comment:
Every Lie ideal is a subalgebra. I guess this might be a useful `def` or `instance`.